---
title: "600 cell analysis"
output: html_notebook
---

```{r Set up environment}
#Define packages for install and/or library loading 
pkg_list_cran <- c('BiocManager', 'dplyr', 'Matrix', 'ggplot2', 'rmarkdown')
pkg_list_bioc <- c('Seurat', 'SingleCellExperiment', 'scater')
pkg_list_all <- c(pkg_list_cran, pkg_list_bioc)

##############################################################

# ONLY RUN THIS SECTION IF INSTALLING PACKAGES FROM SCRATCH #

#Install packages
#lapply(pkg_list_cran, install.packages)
#lapply(pkg_list_bioc, BiocManager::install)
#install.packages('here')

#Save renv session
#renv::snapshot()

##############################################################

#Restore renv session
renv::restore()

# Update all packages - recommended if it's been a while
# since the code was last run
# renv::update()

#Load libraries
lapply(pkg_list_all, library, character.only=TRUE)

#Set seed for random number generation
set.seed(444)

```

```{r Define paths}

#Set top-level directory using the here package
here::i_am("600_cells_analysis/600_cells_analysis.Rmd")
library(here)
here()

#Select experiment - CHANGE THIS EACH TIME
exp <- '310523_PR0979_v2'

#Set paths to save and read Seurat objects
ifelse(!dir.exists(here(exp, "r_objects")), dir.create(here(exp, "r_objects")), "Folder exists already")
data_path <- here(exp, 'r_objects')

#Make and/or set path to save figures
ifelse(!dir.exists(here(exp, "r_figures")), dir.create(here(exp, "r_figures")), "Folder exists already")
fig_path <- here(exp, 'r_figures')
```

```{r Convenience functions}
#ReadParseBio with correct count_matrix name
ReadParseBio_count <- function(data.dir, ...) {
  mtx <- file.path(data.dir, "count_matrix.mtx")
  cells <- file.path(data.dir, "cell_metadata.csv")
  features <- file.path(data.dir, "all_genes.csv")
  return(ReadMtx(
    mtx = mtx,
    cells = cells,
    features = features,
    cell.column = 1,
    feature.column = 2,
    cell.sep = ",",
    feature.sep = ",",
    skip.cell = 1,
    skip.feature = 1,
    mtx.transpose = TRUE
  ))
}

#Save figures
SaveFigure <- function(plots, name, type = "png", width, height, res){
  if(type == "png") {
    png(paste0(fig_path, '/', name, ".", type),
      width = width, height = height, units = "in", res = 200)
  } else {
    pdf(paste0(fig_path, '/', name, ".", type),
      width = width, height = height)
}
print(plots)
dev.off()
}

#Save Seurat objects
SaveObject <- function(object, name){
  saveRDS(object, here(exp, 'r_objects', paste0(name, ".RDS")))
}

#Read Seurat objects
ReadObject <- function(name){
  readRDS(here(exp, 'r_objects', paste0(name, ".RDS")))
}
```

```{r Read in data}
#Load in count matrix
data_dir <- here(exp, 'all-sample', 'DGE_filtered')
counts <- ReadParseBio_count(data_dir)

# Check to see if empty gene names are present, add name if so.
table(rownames(counts) == "")
rownames(counts)[rownames(counts) == ""] <- "unknown"

# Read in cell meta data
cell_meta <- read.csv(here(exp, 'all-sample', 'DGE_filtered', "/cell_metadata.csv"), row.names = 1)

#Create Seurat object
seurat_obj <- CreateSeuratObject(counts = counts, names.field = 0, meta.data = cell_meta)
```

FROM PARSE BIO:
When we create our Seurat object the plate well numbers (column names in the expression matrix) from the experiment will automatically be assigned to the cell identity slot. In other words, the program assumes this how we want to initially classify our cells. In general, we would like to avoid this behavior so there isn't a bias towards a particular cell class when removing outliers.

```{r Well numbers}
# Setting our initial cell class to a single type, this will change after clustering. 
seurat_obj@meta.data$orig.ident <- factor(rep("seurat_obj", nrow(seurat_obj@meta.data)))
Idents(seurat_obj) <- seurat_obj@meta.data$orig.ident

#Save raw Seurat object
SaveObject(seurat_obj, 'raw_seurat_obj')
#seurat_obj <- ReadObject('raw_seurat_obj')

#Save raw SingleCellExperiment object
sce_obj <- as.SingleCellExperiment(seurat_obj)
SaveObject(sce_obj, 'raw_sce_obj')
```

We need to exclude multiplets from analysis - these are cells with abnormally high transcript of gene counts.

[FROM BIOSTARS](https://www.biostars.org/p/407036/#:~:text=nCount_RNA%20is%20the%20total%20number,a%20doublet%20(or%20multiplet).)

"nFeature_RNA is the number of genes detected in each cell. nCount_RNA is the total number of molecules detected within a cell. Low nFeature_RNA for a cell indicates that it may be dead/dying or an empty droplet. High nCount_RNA and/or nFeature_RNA indicates that the "cell" may in fact be a doublet (or multiplet). In combination with %mitochondrial reads, removing outliers from these groups removes most doublets/dead cells/empty droplets, hence why filtering is a common pre-processing step.

The NormalizeData step is basically just ensuring expression values across cells are on a comparable scale. By default, it will divide counts for each gene by the total counts in the cell, multiply that value for each gene by the scale.factor (10,000 by default), and then natural log-transform them."

The columns added after quickpercellQC can be described as follows:

- sum = nCount_RNA = lib_size = number of molecules in the well

- detected = nFeature_RNA = number of unique genes in the well

- low_lib_size suggests RNA has been lost at some point during library preparation, either due to cell lysis or inefficient cDNA capture and amplification

- low_n_features suggests cell is dead/dying/"empty droplet"

```{r Cell quality control}
#Add quick per cell QC
sce_obj <- scuttle::quickPerCellQC(sce_obj)

#Load stringr for string functions
library(stringr)

#Add mitoDNA proportion
location <- rowRanges(sce_obj)
is.mito <- sapply(names(seqnames(location)), str_detect, '^mt*')
sce_obj <- addPerCellQCMetrics(sce_obj, subsets = list(Mito = is.mito))
qc_df <- perCellQCMetrics(sce_obj, subsets = list(Mito = is.mito))

#Get the number of cells for which %mito is greater than 10
sum(sce_obj$subsets_Mito2_percent > 10)
```

Alternatively, do a quick version of this QC with Seurat:
```{r Seurat QC}
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
plot <- VlnPlot(seurat_obj, pt.size = 0.10,
features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot
SaveFigure(plot, "vln_QC", width = 12, height = 6)


plot1 <- FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(seurat_obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

plot1 + plot2
SaveFigure((plot1 + plot2),"scatter_QC", width = 12, height = 6, res = 200)
```

Now time to find highly variable features for the whole dataset:
```{r Normalise and find highly variable features}
#Normalise
seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize", scale.factor = 10000)

#Find HVFs
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = "vst", nfeatures = 2000)
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(seurat_obj), 10)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(seurat_obj)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)

plot1 + plot2
SaveFigure((plot1 + plot2), "var_features", width = 12, height = 6)
```

```{r Scale and dimensionality reduction}
#Scale 
seurat_obj <- ScaleData(seurat_obj)

#Dimensionality reduction
seurat_obj <- RunPCA(seurat_obj)
SaveObject(seurat_obj, "seurat_obj_after_PCA")

#Load obj
srat <- ReadObject("seurat_obj_after_PCA")

# Examine and visualize PCA results a few different ways
print(srat[["pca"]], dims = 1:5, nfeatures = 5)

#Plot!
plot <- VizDimLoadings(srat, dims = 1:2, reduction = "pca")
SaveFigure(plot, "viz_PCA_loadings", width = 10, height = 8)

plot <- DimPlot(srat, reduction = "pca", group.by = "orig.ident")
SaveFigure(plot, "pc1_2_scatter", width = 8, height = 6)
```

Now let's look at heterogeneity in the dataset with DimHeatmap
```{r DimHeatmap}
#Plot DimHeatmap for PC1
# NB: For large datasets, set cells = 500
plot <- DimHeatmap(srat, dims = 1, balanced = TRUE, fast = FALSE)
SaveFigure(plot, "dim_heatmap1", width = 8, height = 6)

#Plot DimHeatmap for PC2
# NB: For large datasets, set cells = 500
plot <- DimHeatmap(srat, dims = 1:15, balanced = TRUE, fast = FALSE)
SaveFigure(plot, "dim_heatmap1_15", width = 12, height = 18)
```

```{r Determine dimensionality of the dataset}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
srat <- JackStraw(srat, num.replicate = 100)
srat <- ScoreJackStraw(srat, dims = 1:20)
plot <- JackStrawPlot(srat, dim=1:20)
SaveFigure(plot, "PC_JackStraw_plot", width = 8, height = 10)

#Quicker alternative!
plot <- ElbowPlot(srat,ndims = 50)
SaveFigure(plot, "PC_elbow_plot", width = 8, height = 10)
```

Chose the first 9 PCs, now time to cluster:
```{r Clustering}

### NB: NUMBER OF PCs FOR CLUSTERING = NUMBER OF PCs FOR UMAP


#Find SNN
srat <- Seurat:::FindNeighbors(srat, dims=1:30)
#Run Louvain clustering
srat <- FindClusters(srat, resolution = 0.8)
srat <- FindClusters(srat, resolution = 0.6)
srat <- FindClusters(srat, resolution = 1)
srat <- FindClusters(srat, resolution = 1.2)
srat <- FindClusters(srat, resolution = 1.4)

############################################
##### CLUSTREE #######
######################
#
# Run find cluster step with a range of resolutions 
# On the srat obj, it will generate a column for each
renv::install('clustree')
library(clustree)

plot <- clustree(srat)
SaveFigure(plot, "clustree", width = 10, height = 10)

#renv::install('ape')
#library(ape)


```

```{r Run non-linear dimensionality reduction}

### NB: NUMBER OF PCs FOR CLUSTERING = NUMBER OF PCs FOR UMAP

srat <- RunUMAP(srat, dims=1:30)
plot <- DimPlot(srat, reduction = 'umap')
SaveFigure(plot, 'umap_louvain_res_p3', width=9, height=8)

#Save object
SaveObject(srat, "seurat_obj_clustered")
#srat <- ReadObject("seurat_obj_clustered")



```
NOTES FROM MARCELA:
functional annotation
go genome annotation terms
gene ontology resource
good guides & annotation

cluster profiler r package
generates plots & viz for this type of analysis e.g.

refine clusters
cluster stability
ontology analysis

```{r Find cluster markers (DGE)}
#Find markers per cluster
BiocManager::install("MAST")
library('MAST')
#srat <- BuildClusterTree(srat, reorder = T, reorder.numeric = T)
srat_markers <- FindMarkers(srat, test.use = 'MAST', ident.1=1)

c8_ont <- srat_markers[srat_markers$cluster == '8' & srat_markers$p_val_adj <0.05 & srat_markers$avg_log2FC > 0,]

writeLines(c8_ont$gene, 'c8_ontology.csv')

#Find unique markers per cluster
srat_all_markers <- FindAllMarkers(srat, min.pct = 0.25, logfc.threshold = 0.25)
srat_all_markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)

######
# Use FindMarkers, not FindAllMarkers, use MAST method
######

#Plot expression probability distributions across clusters
plot <- VlnPlot(srat, features = c("Otog", "Armc4"), group.by = "tree.ident")
SaveFigure(plot, "vln_exp1", width = 16, height = 8)

#Plot raw counts
plot <- VlnPlot(srat, features = c("Otog", "Armc4"), slot = "counts", log = TRUE)
SaveFigure(plot, "vln_exp2", width = 16, height = 8)
```

```{r Plot top n genes per cluster}
top5 <- srat_markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)
to_plot <- unique(top5$gene)
plot <- DotPlot(srat, features = to_plot, group.by = "tree.ident") + coord_flip()
SaveFigure(plot, "dplot_top5", width = 9, height = 20)
```

```{r Merging clusters and labelling cell types}
# Plotting PBMC markers
markers <- c("Otog", "Armc4", "Mgat4c", "Kcnh7", "mt-Rnr1", "mt-Rnr2", "Coch",
"Tectb", "Sparcl1", "Miat", "Ush2a", "Mbp")

srat_markers <- FeaturePlot(srat, features = markers)
SaveFigure(srat_markers, "srat_markers", width = 40, height = 30)
```

```{r Plot UMAP with labelled clusters}
plot <- DimPlot(srat, reduction = "umap", label = TRUE) + NoAxes() + NoLegend()
SaveFigure(plot, "umap_louvain_nolegend", width = 8, height = 7)
```

```{r gene ontology analysis}
c8_ont <- srat_markers[srat_markers$cluster == '8' & srat_markers$p_val_adj <0.05 & srat_markers$avg_log2FC > 0,]

writeLines(c8_ont$gene, 'c8_ontology.csv')
```


```{r Plot UMAP with descriptive labelled clusters}
#NB: Taken from tutorial, doesn't work for this dataset, revisit at a later date!

#new_ids <- c("HCs", "SCs Sparcl1+", "SCs Sparcl1-", "mt+ myelin+", "mt+ myelin-", "Lateral wall")

#new_id_list <- list(HCs = c(8,9,10), SCsPlus = c(6,7),
#  SCsMinus = 5, MtPlus = 1, MtMinus = 2, LatWall = c(3,4))

#for (i in 1:length(new_id_list)) {
#  ind <- which(srat@meta.data$tree.ident %in% new_id_list[[i]])
#  srat@meta.data$collapsed[ind] <- names(new_id_list)[i]
#}

#srat@meta.data$collapsed <- factor(
#  srat@meta.data$collapsed, levels = names(new_id_list), ordered = TRUE)
#Idents(srat) <- pbmc@meta.data$collapsed

#names(new_ids) <- levels(srat)
#srat <- RenameIdents(srat, new_ids)

#plot <- DimPlot(srat, reduction = "umap", label = TRUE) + NoAxes() + NoLegend()
#SaveFigure(plot, "umap_louvain_nolegend_names", width = 8, height = 7)
```



```{r Trajectory analysis}
BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'lme4', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment', 'batchelor', 'HDF5Array',
                       'terra', 'ggrastr'))

renv::install('devtools')

devtools::install_github('cole-trapnell-lab/monocle3')

library(monocle3)

remotes::install_github('satijalab/seurat-wrappers')


```
